src/Selenium2Driver.php

```diff
@@ -246,6 +249,68 @@ protected static function charToOptions($char, $modifier = null)
         return json_encode($options);
     }
 
+    /**
+     * Serialize execute arguments (containing web elements)
+     *
+     * @see https://w3c.github.io/webdriver/#executing-script
+     *
+     * @param array $args
+     *
+     * @return array
+     */
+    private function serializeExecuteArguments(array $args)
+    {
+        foreach ($args as $k => $v) {
+            if ($v instanceof NodeElement) {
+                if (strpos($v->getXpath(), self::XPATH_WEBDRIVER_ELEMENT_PREFIX) === 0) {
+                    [$elementUrlEncoded, $elementIdEncoded] = explode(
+                        '--',
+                        substr($v->getXpath(), strlen(self::XPATH_WEBDRIVER_ELEMENT_PREFIX)),
+                        2
+                    );
+
+                    $args[$k] = new Element(base64_decode($elementUrlEncoded), base64_decode($elementIdEncoded));
+                } else {
+                    $args[$k] = $this->findElement($v->getXpath());
+                }
+            } elseif (is_array($v)) {
+                $args[$k] = $this->serializeExecuteArguments($v);
+            }
+        }
+
+        return $args;
+    }
+
+    /**
+     * Unserialize execute result (containing web elements)
+     *
+     * @param mixed $data
+     *
+     * @return mixed
+     */
+    private function unserializeExecuteResult($data)
+    {
+        if ($data instanceof Element) {
+            $minkSession = \Closure::bind(function () {
+                return $this->session;
+            }, $this, CoreDriver::class)();
```

Comment:
this is a no-go. The private state of classes from a different package are not covered by BC. And we do have a (stale) work in progress changing the way NodeElement gets instantiated, which is precisely why the current API of CoreDriver ensures that driver classes don't need to handle that instantiation themselves.

Reply:
`Selenium2Driver` is child class of `CoreDriver` class, thus I find this code quite reasonable as this is exactly what the `CoreDriver` class does https://github.com/minkphp/Mink/blob/97123e0048c3c9db56e153152b11a2d9ba643dbb/src/Driver/CoreDriver.php#L102

is this reasoning enough, or you want me to submit a PR to `behat/mink` to change the visibility of `CoreDriver::session` prop to `protected`? or do you have in mind a better solution?

Reply:
We don't want to make it protected. That would force it to be covered by BC, which would prevent us from removing it from behat/mink when we finally do that refactoring (currently, all our drivers are compatible with the in-progress stale refactoring with no changes in them).

Reply:
So what solution do you propose? If public `CoreDriver::find` method is supposed to keep returning `NodeElement`, I belive the session object will be always needed.

What about moving https://github.com/minkphp/Mink/blob/97123e0048c3c9db56e153152b11a2d9ba643dbb/src/Driver/CoreDriver.php#L102 to a protected method `CoreDriver::createNodeElementFromXpath`? This way, the session object can be kept private.

Reply:
Even if there is no support for native xpath generation from JS/Selenium, given the fact how dummily the xpaths are generated in https://github.com/minkphp/MinkSelenium2Driver/blob/086c6a84093702f89123d3ce42c27d091d9dc16d/src/Selenium2Driver.php#L521 , I belive with one more roundtrip to Selenium, we can obtain xpath via short JS code. This will also solve https://github.com/minkphp/MinkSelenium2Driver/pull/352#discussion_r851288614 cleanly and the returned NodeElement will be like any other.

Reply:
why saying this is dummy ? This is a way to produce the xpath for all those elements without doing more roundtrips to selenium.

Reply:
because if any element matching given xpath is prepended, previously created `NodeElement` will match a different DOM element later - is that wanted?

Reply:
completely reworked, see file changes and https://github.com/minkphp/MinkSelenium2Driver/pull/352#discussion_r854002660 discussion